### ISSUE TITLE
Wrapping my head around Pull Request-based flow

### ISSUE THREAD

**Issue by deitch** [weaveworks/tf-controller#653]
I have been trying, via the various tracking issues for branch-based flow and the docs, to understand what the appropriate flow is for PRs with tf-controller.

In "normal" (does that exist?) software process, I open a PR, it is approved, merged to mainline. The PR itself inherently shows me, via the diff and various applied CI pipelines, what the differences will be. Things can slip through, but minimally.

In terraform, the essence is not only the changed code, but the `plan` output. If I outline the ideal terraform PR, I come up with something like this.

1. Open PR with changes, which shows the code diff
2. Flux performs a `terraform plan`, which then is _posted back to the PR_ - the output of this `plan` is essential for approving the PR
3. An approver sees the PR with the diff _and the `plan` output_ and approves the PR
4. The PR is merged into main
5. Flux sees the merge, runs `plan` again, and only runs `apply` if the `plan` is identical to the PR output.

This becomes really important when thinking about things like race conditions and such, e.g. someone applies another PR right before mine. In the case of software, it matters, but less if it touches other areas; in the case of terraform, the impact can be very large (which is one reason why `plan` exists).

I could see the last step changing if, for example, we did an approve process tied into Actions. I have seen the "change the `Terraform` CRD" flow, which is nice in theory, but hard to do in practice at an org. "Now I need to monitor a different one? And make changes that are not subject to PRs? Or are, so now I need all of these additional steps?"

One other piece is the webhook vs polling. Most of the conversations on the issues here have focused on GitHub Webhook, but lots of environments run their k8s cluster behind firewalls that simply won't give Webhook access.

Finally, in the meantime, is there a recommended way to get a flow that is kind of like the above, until the various features are merged in?

Thanks.

**Comment by yitsushi** [comment]
> This becomes really important when thinking about things like race conditions and such, e.g. someone applies another PR right before mine. In the case of software, it matters, but less if it touches other areas; in the case of terraform, the impact can be very large (which is one reason why plan exists).

There is a feature on github that prevents merging if the PR's head is not up-to-date with the base branch. Doesn't it solve this issue? Previously we (not at Weaveworks) used this feature for this exact reason, to prevent applying Terraform and Puppet in a very state from main.

Wit this enforced PR rule addition I think the only case when the plan output can be different if someone changes the environment manually without the terraform controller.

> Flux sees the merge, runs plan again, and only runs apply if the plan is identical to the PR output.

As the two processes are two separate processes, in my opinion it would add an extra complexity on top of everything that's not necessary. That does not mean I don't agree that the issue exists and its importance, but I don't think that's the way to solve it, at least not as a first-response.

**Comment by yitsushi** [comment]
What I mean about additional complexity:
* The Terarform controller has to check if branch-based planner is enabled on that repo
* The Terraform controller has to check if there is a branch-based planner Terraform object or not.
* The branch-based planner terraform object may or may not exists at that point, because the planner is polling and when a PR is merged/closed we delete the resource (garbage collection).
* Even if the terraform object is there, the terraform controller has to check all of them as the controller can see only commits from main, so it has to match the commits hashes on all branch-based planner teraform objects if the latest commit is the same as the main has.

Potentially, I missed some points, or even something can be simplified and dropped from the list, but I think it gives a general overview what this would require to add to the controller.

**Comment by chanwit** [comment]
Thank you, @deitch, for your inputs.

Generally, yes, we expect this feature to reduce complexity and increase the visibility of the Terraform planning process.
The MVP that @yitsushi has been working on moves towards that goal. I'm quite sure we're on the same page of the idea.

For the webhook part, we dropped the webhook implementation in favor of the polling mechanism. Please refer to `RFC-0001` for tech details around this topic.

> In the meantime, as this works through, what is the recommended way of getting a PR-driven flow using tf-controller?

The only workflow that we support right now is the manual mode:
1. Set the `approvePlan` to be empty by modifying the field in your TF YAML file to `spec.approvePlan: ""`.
2. Modify your Terraform (TF) files as needed and push these changes.
3. A new plan will be automatically generated after your push.
4. To view the changes, run the command `kubectl get tf ...`  You'll notice a message prompting you to set `spec.approvePlan: <plan name>`. This message indicates that you should set the approvePlan field to the name of your newly created plan.
5. Create a new branch in your repository using the command `git checkout -b new-branch`.
6. On this new branch, modify your YAML files according to the suggestion in step 4, i.e., set the `approvePlan` to the name of your new plan.
7. Open a Pull Request (PR) with these changes and send it to your team for review and approval.
8. Once your team approves the changes and the PR is merged into the `main` branch, the provision steps will continue, and the changes you've made will be provisioned.

**Comment by deitch** [comment]
BTW, I am enjoying this conversation for how it is pushing me to question my process assumptions. I assume I have a good flow that I want to adapt, but the issue you raised make me question it.

**Comment by deitch** [comment]
> The only workflow that we support right now is the manual mode

I saw the docs and the demo video. My challenges with it are several:

1. It requires modifying the TF and pushing to / merging into main before the `plan` even is run.
2. It assumes humans have access to the cluster to run `kubectl`
3. It requires modifying two repos - this is an annoyance, but probably can be worked with for a while, if needed

Am I right that it has no way to run `plan` on branches now, at least until branch-based comes in?

**Comment by yitsushi** [comment]
> I read your response, and started thinking down a different path. What if the tf plan during PR were saved, and once it is merged - which in itself means the plan output was approved - rather than a straight tf apply, we do tf apply to the PR-linked plan output?

There are two main issues with this:
1. We have the same race condition, but now we pushed the responsibility from github to tf-controller, when it goes around the next iteration of polling PR changes, if two PRs was merged we have to pick one (most likely the first merged one), and the other merged pr will be lost, what happens with that? It's merged, but will not be applied, the output changes will show up in the next PR, so that PR has a lot of unrelated changes in the output.

2. The core concept of Flux and the world around it is that the reconciled objects are the same as the source says. If the main Terraform object was reconciled BUT the stored commit does not match with the applied plan, it's drifting away and from that point, we can't say we have the resources in sync with the git repository.

> This has to do with the internal of tf-controller, as well as the architecture for branch-based

Not necessarily, Flux will reconcile the Source repository unless you specifically tells Flux to suspend. The Terraform controller's job is to make sure it's updated based on the received git commit hash, plan and apply if possible and it has zero knowledge about other branches or pull requests.

With this logic the Flux-friendly outcome would be simply reject all commits and don't apply any of the commits at all, throw an error during the reconciliation loop and let main sync up with a loop-up PR/commit. It's not reasonable (in the Flux ecosystem) to apply commit `AAAA` if the reconciled source points to commit `BBBB` and say "yes we are in sync, reconciliation completed".

**Comment by deitch** [comment]
> We have the same race condition, but now we pushed the responsibility from github to tf-controller, when it goes around the next iteration of polling PR changes, if two PRs was merged we have to pick one (most likely the first merged one), and the other merged pr will be lost, what happens with that? It's merged, but will not be applied, the output changes will show up in the next PR, so that PR has a lot of unrelated changes in the output

What you are saying is, do not let the second one merge until the first one is merged and applied _and_ the second one has been rebased and approved with a plan post-first merge.

The "rebase" part is a GitHub feature, as described above; the second probably can be done, but as you describe, it is complex, and depends heavily on the user making output from tf-controller valid and in.

I think it would be great if that were easily doable, but "easily" is the obvious challenge.

1. I open a PR
2. It cannot be merged until approved (of course), but also until: a) it is on the latest `main` commit; b) tf-controller has run a plan which I see and approve.

The second PR would also have that, but as soon as the first is merged in, the second now is out of date, requiring a rebase, which means a new `plan`, etc.

> The core concept of Flux and the world around it is that the reconciled objects are the same as the source says. If the main Terraform object was reconciled BUT the stored commit does not match with the applied plan, it's drifting away and from that point, we can't say we have the resources in sync with the git repository

I got lost here

**Comment by yitsushi** [comment]
> The second PR would also have that, but as soon as the first is merged in, the second now is out of date, requiring a rebase, which means a new plan, etc.

Basically yes :smile: 

> I got lost here

I was trying to say...

1. PR-1 merged with `Commit AAAA`
2. PR-2 merged with `Commit BBBB`
1. Flux reconciles the Source Repository to `Commit BBBB`
3. Terraform controller applies `Commit AAAA` and says "yes I'm up-to-date with the source"

That feels wrong. In my opinion, it should either apply `Commit BBBB`, or neither of them and throw an error.

**Comment by deitch** [comment]
> but as soon as the first is merged in, the second now is out of date, requiring a rebase, which means a new plan, etc

That really should be the job of GitHub rules to prevent, right? "Don't let this merge until prerequisites A,B,C,D have been satisfied" is the essence of those rules.

> PR-1 merged with Commit AAAA
> PR-2 merged with Commit BBBB

Ideally we just prevent PR-2 from merging until AAAA is merged, _and_ PR-2 has met all requirements, which _includes_ "rebase to master" _and_ "tf-controller ran the appropriate `plan` and shared its results on the PR". Then the race condition could not happen.

 just don't know how to sync those.

> Flux reconciles the Source Repository to Commit BBBB
> Terraform controller applies Commit AAAA and says "yes I'm up-to-date with the source"

This is the part I didn't understand. Why would Flux and tf-controller be out of sync?

**Comment by yitsushi** [comment]
If we apply a plan from a branch based planner that does not match with the latest commit in the repository, we are out of sync. If they are the same, well we have no issues at all.

In my example `Commit AAAA` and `Commit BBBB` are not depend on each other, they are just two PRs, if we have (as I suggested as a solution) the github check is enabled to ensure the pr uses the latest commit from the base branch, I don't think we have any issues, because the PR changed, re-plan, comment the plan back, approve, merge.

The only issue I can see in this case is human negligence, merging a PR without reviewing the new plan. It can be still forced by enabling the github feature that forces a new review if the PR changed.

**Comment by deitch** [comment]
> In my example Commit AAAA and Commit BBBB are not depend on each other, they are just two PRs, if we have (as I suggested as a solution) the github check is enabled to ensure the pr uses the latest commit from the base branch, I don't think we have any issues, because the PR changed, re-plan, comment the plan back, approve, merge.

I get it. Because this check is entirely on the GH side, the tf-controller has no way to know if it is enabled or not (or enforce it), and thus could end up in that invalid situation.

> The only issue I can see in this case is human negligence

Which seems to be about 99.99% of all problems? 😁 

Thanks for expelling it all. I quite appreciate it.

**Comment by chanwit** [comment]
@deitch I'll close this issue for now. Please feel free to reopen it if you have further questions about the branch planner.

Thanks!
